homeassistant/components/fireservicerota/sensor.py

```diff
@@ -0,0 +1,126 @@
+
+
+async def async_setup_entry(
+    hass: HomeAssistantType, entry: ConfigEntry, async_add_entities
+) -> None:
+    """Set up FireServiceRota sensor based on a config entry."""
+    coordinator = hass.data[FIRESERVICEROTA_DOMAIN][entry.entry_id]
+
+    async_add_entities([IncidentsSensor(coordinator, entry)])
+
+
+class IncidentsSensor(RestoreEntity, Entity):
+    """Representation of FireServiceRota incidents sensor."""
+
+    def __init__(self, coordinator, entry):
+        """Initialize."""
+        self._coordinator = coordinator
+        self._entry_id = entry.entry_id
+        self._unique_id = entry.unique_id
+
+        self._state = None
+        self._state_attributes = {}
+
+    @property
+    def name(self) -> str:
+        """Return the name of the sensor."""
+        return "Incidents"
+
+    @property
+    def icon(self) -> str:
+        """Return the icon to use in the frontend."""
+        if "prio" in self._state_attributes:
+            if self._state_attributes["prio"][0] == "a":
+                return "mdi:ambulance"
+
+        return "mdi:fire-truck"
+
+    @property
+    def state(self) -> str:
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def unique_id(self) -> str:
+        """Return the unique ID for this sensor."""
+        return f"{self._unique_id}_Incidents"
+
+    @property
+    def should_poll(self) -> bool:
+        """No polling needed."""
+        return False
+
+    @property
+    def device_state_attributes(self) -> object:
+        """Return available attributes for sensor."""
+        attr = {}
+        data = self._state_attributes
+
+        if data:
+            for value in (
+                "trigger",
+                "created_at",
+                "message_to_speech_url",
+                "prio",
+                "type",
+                "responder_mode",
+                "can_respond_until",
+            ):
+                if data.get(value):
+                    attr[value] = data[value]
+
+            if "address" in data:
+                for address_value in (
+                    "latitude",
+                    "longitude",
+                    "address_type",
+                    "formatted_address",
+                ):
+                    if address_value in data["address"]:
+                        attr[address_value] = data["address"][address_value]
+
+            return attr
+
+    async def async_added_to_hass(self) -> None:
+        """Handle entity which will be added."""
+        await super().async_added_to_hass()
+
+        state = await self.async_get_last_state()
+        if state:
+            self._state = state.state
+            self._state_attributes = state.attributes
+            _LOGGER.debug("Restored entity 'Incidents' state to: %s", self._state)
+
+        self.async_on_remove(self._coordinator.stop_listener)
+
+        self.async_on_remove(
+            async_dispatcher_connect(
```

Comment:
The downside of using the dispatcher is that your UpdateCoordinator is unable to keep track of anyone is listening and so if it should fetch data or not.

Reply:
I have implemented a UpdateCoordinator for the binary_sensor now, this just fetches the Duty calender state every minute. The sensor is fed the websocket data via on_incident callback so no UpdateCoordinator needed for that, only the switch is listening to the dispatcher, when an event is received he fetched  the latest response data, and updates it's state, same when switch is toggled, so I think there is no need for another UpdateCoordinator. Tried to use a Coordinator for this as well but can't call manually update routine from (sync) on_incident() and converting the websocket-client to async just for this is too much right now (it's very lightweight and stable as it is)